From 113cfef3ae961c7986c5e46bb0b7d42ed1e70ae0 Mon Sep 17 00:00:00 2001
From: Jeff Mahoney <jeffm@suse.com>
Date: Tue, 16 Apr 2019 11:12:05 -0400
Subject: [PATCH 14/24] target: add delete_target

The forthcoming python target will allow the user to register a target
implementation in Python.  Since these are not static targets, it would
make sense to allow the target to be unregistered from gdb.

This commit adds a new delete_target() call that removes the target
from the target factories and removes the target subcommand to access it.
The target itself will be cleaned up when the last reference to it is dropped.
---
 gdb/target.c | 17 +++++++++++++++++
 gdb/target.h |  3 +++
 2 files changed, 20 insertions(+)

diff --git a/gdb/target.c b/gdb/target.c
index e3e30afd7a5..f33df728b28 100644
--- a/gdb/target.c
+++ b/gdb/target.c
@@ -287,6 +287,23 @@ information on the arguments for a particular protocol, type\n\
     set_cmd_completer (c, completer);
 }
 
+void
+delete_target (const target_info &t, target_open_ftype *func)
+{
+  struct cmd_list_element *c = NULL;
+  auto &func_slot = target_factories[&t];
+  if (target_factories[&t] != func)
+    {
+      internal_error (__FILE__, __LINE__,
+		      _("target to be deleted was not added (\"%s\")."),
+		      t.shortname);
+    }
+  func_slot = nullptr;
+
+  c = add_alias_cmd (t.shortname, c, no_class, 0, &targetlist);
+  gdb_assert (c == NULL);
+}
+
 /* See target.h.  */
 
 void
diff --git a/gdb/target.h b/gdb/target.h
index 8e51516feab..5da972bdaac 100644
--- a/gdb/target.h
+++ b/gdb/target.h
@@ -2330,6 +2330,9 @@ extern void add_target (const target_info &info,
 			target_open_ftype *func,
 			completer_ftype *completer = NULL);
 
+
+extern void delete_target (const target_info &t, target_open_ftype *func);
+
 /* Adds a command ALIAS for the target described by INFO and marks it
    deprecated.  This is useful for maintaining backwards compatibility
    when renaming targets.  */
-- 
2.26.2

